File: enoch_williams/project04/Primes.java
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

public class Primes {

    public static boolean isPrime(long n) {
        if (n <= 1) return false;
        if (n <= 3) return true;
        if (n % 2 == 0 || n % 3 == 0) return false;
        long limit = (long) Math.sqrt((double) n);
        for (long i = 5; i <= limit; i += 6) {
            if (n % i == 0 || n % (i + 2) == 0) return false;
        }
        return true;
    }

    public static List<Long> readNumbers(String path) throws IOException {
        File f = new File(path);
        BufferedReader br = new BufferedReader(new FileReader(f));
        List<Long> nums = new ArrayList<>();
        String line;
        while ((line = br.readLine()) != null) {
            line = line.trim();
            if (line.isEmpty()) continue;
            try {
                long v = Long.parseLong(line);
                nums.add(v);
            } catch (NumberFormatException e) {
                // skip
            }
        }
        br.close();
        return nums;
    }

    public static void main(String[] args) {
        if (args.length < 1) {
            System.out.println("Usage: java Primes <file>");
            System.exit(1);
        }
        String path = args[0];
        List<Long> nums;
        try {
            nums = readNumbers(path);
        } catch (IOException e) {
            System.out.println("Failed to read file: " + e.getMessage());
            return;
        }

        System.out.printf("File: %s (%d numbers)\n\n", path, nums.size());

        long t1 = System.nanoTime();
        long singleCount = 0;
        for (long v : nums) if (isPrime(v)) singleCount++;
        long t1e = System.nanoTime() - t1;

        System.out.println("[Single-Threaded]");
        System.out.printf("  Primes found: %d\n", singleCount);
        System.out.printf("  Time: %.3f ms\n\n", t1e / 1e6);

        int threads = Runtime.getRuntime().availableProcessors();
        int chunk = (nums.size() + threads - 1) / threads;
        ExecutorService pool = Executors.newFixedThreadPool(threads);
        List<Callable<Long>> tasks = new ArrayList<>();
        for (int i = 0; i < threads; i++) {
            final int lo = i * chunk;
            int hi = Math.min(lo + chunk, nums.size());
            if (lo >= nums.size()) break;
            final List<Long> slice = nums.subList(lo, hi);
            tasks.add(() -> {
                long local = 0;
                for (long v : slice) if (isPrime(v)) local++;
                return local;
            });
        }

        long t2 = System.nanoTime();
        long total = 0;
        try {
            List<Future<Long>> futures = pool.invokeAll(tasks);
            for (Future<Long> f : futures) total += f.get();
        } catch (InterruptedException | ExecutionException e) {
            System.out.println("Execution error: " + e.getMessage());
            pool.shutdownNow();
            return;
        }
        long t2e = System.nanoTime() - t2;
        pool.shutdown();

        System.out.printf("[Multi-Threaded] (%d threads)\n", threads);
        System.out.printf("  Primes found: %d\n", total);
        System.out.printf("  Time: %.3f ms\n\n", t2e / 1e6);

        double speedup = (double) t1e / (double) t2e;
        System.out.printf("Speedup: %.2fx\n", speedup);
    }
}

File: enoch_williams/project04/Primes.kt
import java.io.File
import java.util.concurrent.Callable
import java.util.concurrent.Executors
import kotlin.math.sqrt

fun isPrime(n: Long): Boolean {
    if (n <= 1L) return false
    if (n <= 3L) return true
    if (n % 2L == 0L || n % 3L == 0L) return false
    val limit = kotlin.math.sqrt(n.toDouble()).toLong()
    var i = 5L
    while (i <= limit) {
        if (n % i == 0L || n % (i + 2) == 0L) return false
        i += 6
    }
    return true
}

fun readNumbers(path: String): List<Long> {
    val file = File(path)
    val list = ArrayList<Long>()
    file.forEachLine { line ->
        val s = line.trim()
        if (s.isEmpty()) return@forEachLine
        val v = s.toLongOrNull()
        if (v != null) list.add(v)
    }
    return list
}

fun main(args: Array<String>) {
    if (args.isEmpty()) {
        println("Usage: kotlin Primes.kt <file>")
        return
    }
    val path = args[0]
    val nums = try {
        readNumbers(path)
    } catch (e: Exception) {
        println("Failed to read file: ${e.message}")
        return
    }

    println("File: $path (${nums.size} numbers)\n")

    val t1 = System.nanoTime()
    var singleCount = 0L
    for (v in nums) if (isPrime(v)) singleCount++
    val t1e = System.nanoTime() - t1

    println("[Single-Threaded]")
    println("  Primes found: $singleCount")
    println("  Time: ${t1e / 1e6} ms\n")

    val threads = Runtime.getRuntime().availableProcessors()
    val chunk = (nums.size + threads - 1) / threads
    val pool = Executors.newFixedThreadPool(threads)
    val tasks = ArrayList<Callable<Long>>()
    var i = 0
    while (i < threads) {
        val lo = i * chunk
        var hi = lo + chunk
        if (lo >= nums.size) break
        if (hi > nums.size) hi = nums.size
        val slice = nums.subList(lo, hi)
        tasks.add(Callable {
            var local = 0L
            for (v in slice) if (isPrime(v)) local++
            local
        })
        i++
    }

    val t2 = System.nanoTime()
    val futures = pool.invokeAll(tasks)
    var total = 0L
    for (f in futures) total += f.get()
    val t2e = System.nanoTime() - t2
    pool.shutdown()

    println("[Multi-Threaded] ($threads threads)")
    println("  Primes found: $total")
    println("  Time: ${t2e / 1e6} ms\n")

    val speedup = t1e.toDouble() / t2e.toDouble()
    println("Speedup: ${"%.2f".format(speedup)}x")
}
